system/lib/wasmfs/proxied_async_js_impl_backend.h

```diff
@@ -0,0 +1,192 @@
+
+#include "async_callback.h"
+#include "backend.h"
+#include "thread_utils.h"
+#include "wasmfs.h"
+
+//
+// Similar to JSImplBackend, but proxies to another thread where the JS can be
+// async.
+//
+
+using js_index_t = uint32_t;
+
+// A callback state that also adds the C++ component for resuming. This is only
+// needed on the C++ side, but needs to be passed around as part of this
+// struct when we go through C and JS.
+struct CppCallbackState : public CallbackState {
+  // The function to call to resume execution.
+  emscripten::SyncToAsync::Callback resume;
+};
+
+extern "C" {
+// JSImpl async API.
+
+// An async callback with no extra parameters.
+typedef void (*async_callback_t)(CppCallbackState* state);
+
+void _wasmfs_jsimpl_async_alloc_file(js_index_t backend,
+                                     js_index_t index,
+                                     async_callback_t callback,
+                                     CppCallbackState* state);
+void _wasmfs_jsimpl_async_free_file(js_index_t backend,
+                                    js_index_t index,
+                                    async_callback_t callback,
+                                    CppCallbackState* state);
+void _wasmfs_jsimpl_async_write(js_index_t backend,
+                                js_index_t index,
+                                const uint8_t* buffer,
+                                size_t length,
+                                off_t offset,
+                                async_callback_t callback,
+                                CppCallbackState* state);
+void _wasmfs_jsimpl_async_read(js_index_t backend,
+                               js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset,
+                               async_callback_t callback,
+                               CppCallbackState* state);
+void _wasmfs_jsimpl_async_get_size(js_index_t backend,
+                                   js_index_t index,
+                                   async_callback_t callback,
+                                   CppCallbackState* state);
+}
+
+namespace wasmfs {
+
+class ProxiedAsyncJSImplFile : public DataFile {
+  emscripten::SyncToAsync& proxy;
+
+  js_index_t getBackendIndex() {
+    static_assert(sizeof(backend_t) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(getBackend());
+  }
+
+  js_index_t getFileIndex() {
+    static_assert(sizeof(this) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(this);
+  }
+
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_write(
+        getBackendIndex(),
+        getFileIndex(),
+        buf,
+        len,
+        offset,
+        [](CppCallbackState* state) { (*state->resume)(); },
```

Comment:
I don't believe there's any need to dereference the function pointer before calling it.
```suggestion
        [](CppCallbackState* state) { state->resume(); },
```